From bba700889069dfff8afef3959bc0a4c3411c638f Mon Sep 17 00:00:00 2001
From: Lucas Colley <lucas.colley8@gmail.com>
Date: Mon, 22 Dec 2025 16:18:19 +0000
Subject: [PATCH] BLD: linalg: `link_language: 'fortran'` for `_fblas` (#24223)

---
 scipy/linalg/meson.build | 1 +
 1 file changed, 1 insertion(+)

diff --git a/scipy/linalg/meson.build b/scipy/linalg/meson.build
index 5bd12a4436..397a1a5ff3 100644
--- a/scipy/linalg/meson.build
+++ b/scipy/linalg/meson.build
@@ -69,6 +69,7 @@ py3.extension_module('_fblas',
   link_args: version_link_args,
   dependencies: [lapack_lp64_dep, fortranobject_dep],
   install: true,
+  link_language: 'fortran',
   subdir: 'scipy/linalg'
 )
 
