From 2a49f74cf35801ca4547da47cdb5c3cc1815e512 Mon Sep 17 00:00:00 2001
From: "H. Vetinari" <h.vetinari@gmx.com>
Date: Thu, 28 Sep 2023 20:13:37 +1100
Subject: [PATCH 1/2] remove comment that breaks parsing

---
 pyproject.toml | 2 --
 1 file changed, 2 deletions(-)

diff --git a/pyproject.toml b/pyproject.toml
index 95d0ccbe9..851fe8a92 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -12,8 +12,6 @@ build-backend = 'mesonpy'
 requires = [
     # Reason for `<`: future-proofing (0.14.0 released and known to work)
     "meson-python>=0.12.1,<0.15.0",
-    # Reason for `<`: we want to build the 1.11.x releases with Cython 0.29.x
-    # (too many changes in 3.0)
     "Cython>=0.29.35,<3.0",   # when updating version, also update check in meson.build
     # Reason for `<`: future-proofing (2.11.1 is the most recent version)
     "pybind11>=2.10.4,<2.11.1",
